examples/consistency_models/train_consistency_distillation.py

```diff
@@ -0,0 +1,724 @@
+    else:
+        datasets.utils.logging.set_verbosity_error()
+        diffusers.utils.logging.set_verbosity_error()
+
+    # Handle the repository creation
+    if accelerator.is_main_process:
+        if args.push_to_hub:
+            if args.hub_model_id is None:
+                repo_name = get_full_repo_name(Path(args.output_dir).name, token=args.hub_token)
+            else:
+                repo_name = args.hub_model_id
+            create_repo(repo_name, exist_ok=True, token=args.hub_token)
+            repo = Repository(args.output_dir, clone_from=repo_name, token=args.hub_token)
+
+            with open(os.path.join(args.output_dir, ".gitignore"), "w+") as gitignore:
+                if "step_*" not in gitignore:
+                    gitignore.write("step_*\n")
+                if "epoch_*" not in gitignore:
+                    gitignore.write("epoch_*\n")
+        elif args.output_dir is not None:
+            os.makedirs(args.output_dir, exist_ok=True)
+
+    # Initialize the model, using a smaller model than the one defined in the original paper by default
+    if args.model_config_name_or_path is None:
+        model = UNet2DModel(
+                sample_size= args.resolution,
+                in_channels=3,
+                out_channels=3,
+                layers_per_block=2,
+                num_class_embeds=1000,
+                block_out_channels= [32, 64],
+                attention_head_dim=8,
+                down_block_types= [
+                    "ResnetDownsampleBlock2D",
+                    "AttnDownBlock2D",
+                ],
+                up_block_types= [
+                    "AttnUpBlock2D",
+                    "ResnetUpsampleBlock2D",
+                ],
+                resnet_time_scale_shift="scale_shift",
+                upsample_type="resnet",
+                downsample_type="resnet"
+                )
+        target_model = UNet2DModel(
+                sample_size= args.resolution,
+                in_channels=3,
+                out_channels=3,
+                layers_per_block=2,
+                num_class_embeds=1000,
+                block_out_channels= [32, 64],
+                attention_head_dim=8,
+                down_block_types= [
+                    "ResnetDownsampleBlock2D",
+                    "AttnDownBlock2D",
+                ],
+                up_block_types= [
+                    "AttnUpBlock2D",
+                    "ResnetUpsampleBlock2D",
+                ],
+                resnet_time_scale_shift="scale_shift",
+                upsample_type="resnet",
+                downsample_type="resnet"
+                )
+    else:
+        config = UNet2DModel.load_config(args.model_config_name_or_path)
+        model = UNet2DModel.from_config(config)
+        target_model = UNet2DModel.from_config(config)
+
+    
+    # load the model to distill into a consistency model
+    teacher_model = DDPMPipeline.from_pretrained("google/ddpm-cifar10-32").unet
```

Comment:
Not clear what `model`, `target_model`, and `teacher_model` mean.

Reply:
My understanding is as follows (things in square brackets [] are the variable names in the current code):

- The `teacher_model` is the already-trained diffusion model we want to distill to a consistency model (for example, `google/ddpm-cifar10-32`).
- The `model` is a consistency model which is used to calculate the current output of the distilled consistency model $`\boldsymbol{f_\theta}(x_{t_{n + 1}}, t_{n + 1})`$ [`model_output` no. 1*] on the noisy sample $`x_{t_{n + 1}}`$ [`noised_image`] (following Algorithm 2 in the [paper](https://arxiv.org/pdf/2303.01469.pdf)). The parameters $`\boldsymbol{\theta}`$ are updated via gradient descent. (The paper calls this the "online" model, following RL terminology.)
- The `target_model` is a consistency model which is used to calculate the distillation target, which we want the consistency model output $`\boldsymbol{f_\theta}(x_{t_{n + 1}}, t_{n + 1})`$ to match. This is obtained by first running one step of the sampler (in the paper, one step of the Karras Heun sampler) on the `teacher_model` to obtain a previous noisy sample $`\hat{x}_{t_n}^\phi`$ [`model_output` no. 2*], and then getting the `target_model` consistency model's output $`\boldsymbol{f_{\theta^-}}(\hat{x}_{t_{n}}^\phi, t_{n})`$ [`model_output` no. 3*] on $`\hat{x}_{t_n}^\phi`$. The target parameters $`\boldsymbol{\theta^-}`$ are updated via an EMA update (using $`\boldsymbol{\theta}`$). (This is analogous to target models in [DQN-style](https://spinningup.openai.com/en/latest/algorithms/ddpg.html) RL algorithms, and I believe the EMA update is equivalent to what is sometimes called polyak averaging in RL.)

The loss function $`L^{\phi}(\boldsymbol{\theta}, \boldsymbol{\theta^-})`$ is then calculated using a distance metric $`d(\cdot, \cdot)`$ (MSE in the code):

```math
L^{\phi}(\boldsymbol{\theta}, \boldsymbol{\theta^-}) = d(\boldsymbol{f_\theta}(x_{t_{n + 1}}, t_{n + 1}), \boldsymbol{f_{\theta^-}}(\hat{x}_{t_{n}}^\phi, t_{n}))
```
and then we take a gradient descent step with respect to `model`'s parameters $`\boldsymbol{\theta}`$. (It's not clear to me which of `model` and `target_model` is outputted as the final trained checkpoint in the end; according to the paper, theoretically the parameters for the models should converge during training [e.g. $`\boldsymbol{\theta} = \boldsymbol{\theta^-}`$ at the end of training], so using either one should be fine.)

Some proposals:

- I agree that what `model`, `target_model`, `teacher_model` are doing might be hard to follow from the code alone, perhaps it might make sense to add comments explaining what each one does when they are initialized?
- Rename `model_output` no. 2 to `teacher_model_output` and `model_output` no. 3 to `target_model_output`. (If we want to rename `model` [see below], perhaps `model_output` no. 1 can be renamed similarly.)
- Perhaps renaming `model` to something like `online_model` or `current_consistency_model` might be more clear? (Not sure on this one, especially the `current_consistency_model` suggestion, since I'm not sure if it's accurate; would feel more comfortable with this if `model` is what is outputted as the final trained checkpoint, see above.)

(*) Technically, these quantities correspond to `distiller`, `denoised_image`, and `distiller_target` in the code, respectively, but they're all preceded by a calculated raw U-Net output named `model_output`.

Reply:
Thanks so much for explaining @dg845. See my responses below:

> I agree that what model, target_model, teacher_model are doing might be hard to follow from the code alone, perhaps it might make sense to add comments explaining what each one does when they are initialized?

100% agree. The math is heavy enough and we should grab the opportunity here to make it easily digestible to the end users. 

> Rename model_output no. 2 to teacher_model_output and model_output no. 3 to target_model_output. 

Works for me. 

> Perhaps renaming model to something like online_model or current_consistency_model might be more clear? (Not sure on this one, especially the current_consistency_model suggestion, since I'm not sure if it's accurate; would feel more comfortable with this if model is what is outputted as the final trained checkpoint, see above.)

Let's maybe try to output EMA'd model only since that is more common in practice?


> This is obtained by first running one step of the sampler (in the paper and code, one step of the Karras Heun sampler) on the teacher_model to obtain a previous noisy sample 
 [model_output no. 2*]

Have we checked if the `teacher_model` performs okay with that sampler? Does it matter?

Reply:
Sounds like a plan! Thanks so much for being generous with your explanations!